src/round.rs

```diff
@@ -0,0 +1,327 @@
+	votes: HashMap<Id, (Vote, Signature)>,
+	current_weight: usize,
+}
+
+impl<Id: Hash + Eq + Clone, Vote: Clone + Eq, Signature: Clone> VoteTracker<Id, Vote, Signature> {
+	fn new() -> Self {
+		VoteTracker {
+			votes: HashMap::new(),
+			current_weight: 0,
+		}
+	}
+
+	// track a vote. if the vote is an equivocation, returns a proof-of-equivocation and
+	// otherwise notes the current amount of weight on the tracked vote-set.
+	//
+	// since this struct doesn't track the round number of votes, that must be set
+	// by the caller.
+	fn add_vote(&mut self, id: Id, vote: Vote, signature: Signature, weight: usize)
+		-> Result<(), Equivocation<Id, Vote, Signature>>
+	{
+		match self.votes.entry(id.clone()) {
+			Entry::Vacant(mut vacant) => {
+				vacant.insert((vote, signature));
+			}
+			Entry::Occupied(mut occupied) => {
+				if occupied.get().0 != vote {
+					return Err(Equivocation {
+						round_number: 0,
+						identity: id,
+						first: occupied.get().clone(),
+						second: (vote, signature),
+					})
+				}
+			}
+		}
+
+		self.current_weight += weight;
+
+		Ok(())
+	}
+}
+
+/// Parameters for starting a round.
+pub struct RoundParams<Id: Hash + Eq, H> {
+	/// The round number for votes.
+	pub round_number: u64,
+	/// Actors and weights in the round.
+	pub voters: HashMap<Id, usize>,
+	/// The base block to build on.
+	pub base: (H, usize),
+}
+
+#[derive(Debug)]
+pub enum Error<Id, H, S> {
+	PrevoteEquivocation(Equivocation<Id, Prevote<H>, S>),
+	PrecommitEquivocation(Equivocation<Id, Precommit<H>, S>),
+	Chain(::Error),
+}
+
+/// Stores data for a round.
+pub struct Round<Id: Hash + Eq, H: Hash + Eq, Signature> {
+	graph: VoteGraph<H, VoteCount>,
+	prevote: VoteTracker<Id, Prevote<H>, Signature>,
+	precommit: VoteTracker<Id, Precommit<H>, Signature>,
+	round_number: u64,
+	voters: HashMap<Id, usize>,
+	faulty_weight: usize,
+	total_weight: usize,
+	prevote_ghost: Option<(H, usize)>,
+	estimate: Option<(H, usize)>,
+	completable: bool,
+}
+
+impl<Id: Hash + Clone + Eq, H: Hash + Clone + Eq + Ord, Signature: Eq + Clone> Round<Id, H, Signature> {
+	/// Create a new round accumulator for given round number and with given weight.
+	/// Not guaranteed to work correctly unless total_weight more than 3x larger than faulty_weight
+	pub fn new(round_params: RoundParams<Id, H>) -> Self {
+		let (base_hash, base_number) = round_params.base;
+		let total_weight: usize = round_params.voters.values().cloned().sum();
+		let faulty_weight = total_weight.saturating_sub(1) / 3;
+
+		Round {
+			round_number: round_params.round_number,
+			faulty_weight: faulty_weight,
+			total_weight: total_weight,
+			voters: round_params.voters,
+			graph: VoteGraph::new(base_hash, base_number),
+			prevote: VoteTracker::new(),
+			precommit: VoteTracker::new(),
+			prevote_ghost: None,
+			estimate: None,
+			completable: false,
+		}
+	}
+
+	/// Import a prevote. Has no effect on internal state if an equivocation.
```

Comment:
This also means that we need to a +self.faulty_weight in the expression for remaining_commit_votes since if we saw v vote for B_2, that doesn't rule out there being a vote for B_1 which could be used to finalise a block.